hare/broker.go

```diff
@@ -168,8 +170,14 @@ func (b *Broker) HandleMessage(ctx context.Context, pid peer.ID, msg []byte) pub
 }
 
 func (b *Broker) queueMessage(ctx context.Context, pid p2p.Peer, msg []byte) (*msgRPC, error) {
+	b.queueMessageWgMu.Lock()
```

Comment:
i doubt adding mutex is the right thing to do. sync.Waitgroup is threadsafe. the general problem is that add() is called concurrently with wait(). i.e. wait() is executed before the counter increase in a separate goroutine.

see examples in https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/sync/waitgroup_test.go

Reply:
@countvonzero 
I also had some thoughts about this line.

Yes, it's thread-safe if `Add` and `Wait` aren't called concurrently. However, `libp2p` controls the spawning of goroutines that call `queueMessage`. E.g. such a goroutine can be spawned in https://github.com/libp2p/go-libp2p-pubsub/blob/master/validation.go#L407. 

How would you suggest to rewrite this code?

Reply:
> Yes, it's thread-safe if Add and Wait aren't called concurrently.

it's thread-safe, period. Add and Wait can be called concurrently. the only restrictions is, in go doc https://pkg.go.dev/sync#WaitGroup,  
```
Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.
```
i.e. at least one `Add` needs to be called before a `Wait`.
and it's hard for me to imagine that our code call Close (that calls Wait) before the p2p layer called HandleMessage (that calls queueMessage)?

but first thing first. it's still unclear to me what the data race is. i cannot reproduce this. my cmds are
```
go clean -testcache && go test -race -v ./hare -run Test_multipleCPsAndIterations -count 10
```
and the issue referenced didn't seem to be directly about this PR. what am i missing? how do i reproduce this?

Reply:
> and it's hard for me to imagine that our code call Close (that calls Wait) before the p2p layer called HandleMessage (that calls queueMessage)?

could be the case in some tests where Close is called when the counter is 0, and after that some message is received. i would get rid of queueMessage anyway, as it definitely not correct if goroutine is created by pubsub. instead, just use a `closed` variable to exit goroutine early

Reply:
@countvonzero 

> it's thread-safe, period. Add and Wait can be called concurrently. the only restrictions is, in go doc https://pkg.go.dev/sync#WaitGroup,

Well, by `Add` I meant `Add` with a positive delta because `Done` is usually used instead of `Add` with a negative delta.

> it's hard for me to imagine that our code call Close (that calls Wait) before the p2p layer called HandleMessage (that calls queueMessage)?

As Dmitry said, it can happen (and it sometimes happens) in tests.

> it's still unclear to me what the data race is.

The data race is concurrent write/read in `sync.(*WaitGroup).Add` in `sync.(*WaitGroup).Wait`. I couldn't find the log because it was seen locally. I didn't create a separate issue for it.

> the issue referenced didn't seem to be directly about this PR

The data races from the issue description happen because of logging after a test is completed, the added `sync.WaitGroup`'s fix those data races.

Reply:
@dshulyak do you mean getting rid of both `queueMessage` and `queueChannel`?

Reply:
>> it's hard for me to imagine that our code call Close (that calls Wait) before the p2p layer called HandleMessage (that calls queueMessage)?
>
> As Dmitry said, it can happen (and it sometimes happens) in tests.

but in tests the p2p layer is mocked out. it doesn't make sense to add a mutex just because the test isn't setup correctly.
and if the broker is shutdown, HandleMessage() should simply check for some closed context and abort, instead of calling `queueMessage`

Reply:
> but in tests the p2p layer is mocked out. it doesn't make sense to add a mutex just because the test isn't setup correctly.

some tests are definitely using in-memory connections. i remember refactoring them when adding libp2p.

> @dshulyak do you mean getting rid of both queueMessage and queueChannel?

no i meant to get rid of wait group. instead introduce a variable (a context or atomic boolean, whatever makes more sense). after Close is called any future queueMessage call should be exited immediately if variable is "true"

Reply:
> but in tests the p2p layer is mocked out. it doesn't make sense to add a mutex just because the test isn't setup correctly.

by this i mean one should be able to stop p2p mocknet before `broker.Close()` is called.

Reply:
unfortunately pubsub router doesn't have a synchronous Close method. it accepts context and will terminate goroutines asynchronously, but there is no way to wait for that